refactor lib npm.js

isaacs opened the pull request:
Cleaning up `lib/npm.js`, `lib/help.js`, `bin/npm-cli.js`, and creating 100% test coverage along the way.

Come and get it!  Packed full of refactoring and shiny new unit tests.

isaacs committed:
unit test for unsupported version checker

Also, modernize the code in the unsupported version checking module.

isaacs committed:
add unit test for cmd-list module

isaacs committed:
remove some unused utils and outdated tests

isaacs committed:
Module to get project scope from prefix

isaacs committed:
module to listen to process log events

isaacs committed:
module for dereferencing aliases to canonical cmd names

isaacs committed:
module to setup log settings based on configs

isaacs committed:
module to delete old debug log files

isaacs committed:
Remove the log-stream config, fix some references to removed modules

isaacs committed:
remove onload-script config

isaacs committed:
Update open-url and help to fix browser help docs

Allows openUrl to open `file://` urls, and passes that instead of the
raw path.

isaacs committed:
add @npmcli/ci-detect dependency

isaacs committed:
remove reference to npm.rollbacks list

We don't use that any more, Arborist handles it all for us.

isaacs committed:
First pass refactoring npm.js and npm-cli.js

- Several sections of code moved out into separate modules.
- npm is an instance of an inline class.
- uses a Proxy for npm.commands object instead of the old hand-rolled
  Proxy-esque thing.
- organized code path for outputting usage/help information.

isaacs committed:
update dependencies, refactor config loading to async

This removes a lot of very outdated dependencies, updates many to
their modern (usually promisified) versions, and updates (or removes)
code to account for the change.

Several dependencies have been completely removed, and others a bit
shuffled around, so that the node_modules folder can be bundled somewhat
more optimally than it would have otherwise.

isaacs committed:
move npm-cache-filename to dev deps

isaacs committed:
module to set user-agent config

isaacs committed:
Assign usage/completion to npm.commands[x] for completion support

isaacs committed:
use set-user-agent util

isaacs committed:
move update-notifier into separate module

isaacs committed:
update chalk, remove is-ci

isaacs committed:
update parse-conflict-json

isaacs committed:
remove direct dep on npm-install-checks

This module is only used via arborist and npm-pick-manifest

isaacs committed:
add tests for is-windows* modules

isaacs committed:
tests for error-message util

isaacs committed:
remove unnecessary is-registry util

isaacs committed:
refactor and test perf.js util

isaacs committed:
protect against non-strings when dereferencing commands

isaacs committed:
code style cleanup in error-handler util

isaacs committed:
add test for output util

isaacs committed:
make ll a first-class command to make the cmd implementation easier

It was a mistake to have a command alias that sets a config.

isaacs committed:
tests for main npm module

isaacs committed:
refactor and test for main cli script

isaacs committed:
make default help banner more helpful

isaacs committed:
disable progress when --silent is set

isaacs committed:
test coverage and bugfix for reify-output util

isaacs committed:
move reify-output util test to proper location

ruyadorno commented:
@isaacs given the new default for `npm ls` I think it would be nice to also set `all: true` for either `npm ll` or `npm la` in order to make them more useful 😊

isaacs commented:
This PR fixes #1506 and #1533.

isaacs committed:
Full dependency reboot

Reinstall everything from a clean node_modules and package-lock.json
state.

Re-generate list of bundleDependencies and node_modules/.gitignore with
a script that does the right thing based on actual dependency state.

isaacs committed:
Do not exit error if install has an audit advisory

isaacs committed:
Make ll completion test pass by comparing toString() values

The functions are different actual things because only one of them is
loaded in the NYC transpiled coverage environment.

isaacs committed:
test: properly clean reify-output snapshot

isaacs committed:
Use eslint directly instead of standard

Way too opinionated and hard to configure.  Copied out standard's config
as a starting point, and tweaked a few things that don't make sense for
the npm cli project.

isaacs committed:
lint the lib folder

isaacs committed:
script to bundle and gitignore deps as appropriate

isaacs committed:
@npmcli/arborist@0.0.5

isaacs committed:
reset and update all deps

isaacs committed:
reset deps using npm v7

First self-install!

isaacs committed:
install: execute scripts on local install with no args

isaacs committed:
update CI test script for travis

isaacs committed:
@npmcli/arborist@0.0.9, fix self-install

isaacs committed:
Pass options to arborist.reify() in dedupe, prune

ruyadorno commented on lib/utils/proc-log-listener.js:
```diff
@@ -0,0 +1,16 @@
+        // if it crashed once, it might again!
+        log.verbose(`attempt to log ${inspect([level, ...args])} crashed`, ex)
+      } catch (ex2) {
+        console.error(`attempt to log ${inspect([level, ...args])} crashed`, ex)
```
```suggestion
        console.error(`attempt to log ${inspect([level, ...args])} crashed`, ex2)
```

ruyadorno reviewed (approved):
LGTM 👍 :shipit:

isaacs committed:
update minipass-pipeline to fix caching bug

isaacs committed:
remove unnecessary package overrides from licensee config

isaacs committed:
Reset deps (again)

This resets the node_modules tree state after fixing the bug in
minipass-pipeline that caused some tarballs to be incompletely unpacked.

isaacs committed:
@npmcli/arborist@0.0.10

isaacs committed:
test: more affordances for Windows
